upsert/loopdetector.go

```diff
@@ -0,0 +1,76 @@
+package upsert
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/bombsimon/logrusr"
+	"github.com/go-logr/logr"
+	"github.com/google/go-cmp/cmp"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// TODO: Should we try to find a way to enable this in CI only?
```

Comment:
If you switch the thing that provides CreateOrUpdate into an interface, I'd prefer to have a field there for this, instead of a package level var. If the only place where we're going to pay attention to it is in CI, I'd say yes it makes sense to only enable in CI.

Reply:
Switched this to an interface. How should we plug "runs in CI" through, add a  `--enable-ci-debug-output` flag to "hypershift install" maybe?

Reply:
nvm, I just made the cmd default it to true when the `CI` env var is `true` so we don't need two changes